docs/user-guide/dpo.rst

```diff
@@ -94,6 +98,63 @@ Always follow the prompt-response template format used during your SFT training
 
+Sequence Packing with DPO
+=========================
+
+We also support packed sequence training with DPO. Sequence packing is a training technique in which multiple training examples are concatenated to create one longer sequence. This approach eliminates the need for padding and improves GPU utilization.
+Refer to the `sequence packing documentation <https://docs.nvidia.com/nemo-framework/user-guide/latest/nemotoolkit/features/optimizations/sequence_packing.html?highlight=packing#>`_ for a detailed overview of sequence packing and its advantages. This document 
+discusses sequence packing for SFT in particular, but the same benefits apply to DPO, as well.
+
+Packing your DPO dataset is done as a preprocessing step in NeMo and NeMo-Aligner. We provide a `script https://github.com/NVIDIA/NeMo-Aligner/blob/ashors/dpo-packing/examples/nlp/data/dpo/prepare_packed_dpo_dataset.py`_ to pack you DPO dataset. This script assumes you alrady have a prepared DPO-format dataset. Three main steps are run in this script:
+
+  #. The online processing code in ``DPOModelDataset`` is run. This includes tasks such as prompt template manipulation and tokenization. The result is an array of tokenized sequences, represented by indices.
+  #. Chosen and rejected sequences are concatenated.
+  #. The tokenized sequences are grouped by length and a packing algorithm is run.
+
+
+You can read more about packing algorithms  `here <https://en.wikipedia.org/wiki/Bin_packing_problem#Offline_algorithms>`_. Currently, two variants of ``first_fit`` are supported:
+
+  #. ``first_fit_decreasing``: sorts the sequences in decreasing order before applying the first-fit algorithm. It generates a more optimal packing, but it tends to keep all short sequences together, which may have an impact for convergence.
+  #. ``first_fit_shuffle``: runs first-fit in a random order. Packing is less optimal but it keeps the dataset order random. The recommendation is to run first_fit_shuffle and check the packed sequence lengths. If they are similar to the target length (i.e. efficient packing), then use shuffle. Otherwise try first_fit_decreasing.
+
+
+The following is an example of running the packing script to prepare your DPO dataset:
+
+.. code-block:: bash
+
+   python examples/nlp/data/dpo/prepare_packed_dpo_dataset.py \
+      model.data.data_prefix=/path/to/training.jsonl \
+      +model.encoder_seq_length=2048 \
+      +tokenizer_path=/path/to/tokenizer/model \
+      +output_dir=/path/to/output_folder \
+      +pack_sizes=[4096] \
+      +tokenizer_type=<huggingface or sentencpiece>
+   [  +packing_algorithm=first_fit_shuffle \  ]
+   [  +seed=0                                 ]
+
+
+Because this script packs chosen and rejected sequences together, ``pack_sizes`` should always be at least double ``model.encoder_seq_length``.
```

Comment:
should it always be double or just recommended? For example, would someone need to double the seqlen if the len(chosen || rej) << encoder_seq_len

Reply:
Depending on the dataset, it might not be required, but it should always be done for datasets which have at least one example where both the chosen and rejected completions are close to `encoder_seq_len`.

Reply:
I see. I guess it's simpler just to say it's double. Thanks for clarifying